contrib/native/client/example/pooledConnections.cpp

```diff
@@ -0,0 +1,300 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <fstream>
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+#include <boost/thread.hpp>
+#include "drill/drillc.hpp"
+
+int nOptions=5;
+
+struct Option{
+    char name[32];
+    char desc[128];
+    bool required;
+}qsOptions[]= {
+    {"query", "Query strings, separated by semicolons", true},
+    {"connectStr", "Connect string", true},
+    {"logLevel", "Logging level [trace|debug|info|warn|error|fatal]", false},
+    {"numConnections", "Number of simultaneous connections", true},
+    {"numIterations", "Number of iterations to run. Each query is sent to each connection this many times", true}
+};
+
+std::map<std::string, std::string> qsOptionValues;
+
+const char* exceptionInject="alter session set `drill.exec.testing.controls` = '{ \"injections\" : [{ \"type\":\"exception\",\"siteClass\":\"org.apache.drill.exec.work.fragment.FragmentExecutor\",\"desc\":\"fragment-execution\",\"nSkip\":0,\"nFire\":1,\"exceptionClass\":\"java.lang.OutOfMemoryError\"}]}'";
+
+Drill::status_t SchemaListener(void* ctx, Drill::FieldDefPtr fields, Drill::DrillClientError* err){
+    if(!err){
+        printf("SCHEMA CHANGE DETECTED:\n");
+        for(size_t i=0; i<fields->size(); i++){
+            std::string name= fields->at(i)->getName();
+            printf("%s\t", name.c_str());
+        }
+        printf("\n");
+        return Drill::QRY_SUCCESS ;
+    }else{
+        std::cerr<< "ERROR: " << err->msg << std::endl;
+        return Drill::QRY_FAILURE;
+    }
+}
+
+boost::mutex listenerMutex;
+Drill::status_t QueryResultsListener(void* ctx, Drill::RecordBatch* b, Drill::DrillClientError* err){
+    boost::lock_guard<boost::mutex> listenerLock(listenerMutex);
+    if(!err){
+        if(b!=NULL){
+            b->print(std::cout, 0); // print all rows
+            std::cout << "DATA RECEIVED ..." << std::endl;
+            delete b; // we're done with this batch, we can delete it
+            return Drill::QRY_FAILURE;
+        }else{
+            std::cout << "Query Complete." << std::endl;
+            return Drill::QRY_SUCCESS;
+		}
+    }else{
+        assert(b==NULL);
+        switch(err->status) {
+            case Drill::QRY_COMPLETED:
+            case Drill::QRY_CANCELED:
+                std::cerr<< "INFO: " << err->msg << std::endl;
+                return Drill::QRY_SUCCESS;
```

Comment:
Yes. Bit of a bad design there. The return value of failure is actually to signal that the listener that the query completed. WE added another api to check the status explicitly later. Unfortunately, the ODBC driver now uses this so hard to change.